apisix/plugins/brotli.lua

```diff
@@ -0,0 +1,195 @@
+            type = "integer",
+            default = 24,
+            enum = {0,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24},
+        },
+        lgblock = {
+            type = "integer",
+            default = 0,
+            enum = {0,16,17,18,19,20,21,22,23,24},
+        },
+        http_version = {
+            enum = {1.1, 1.0},
+            default = 1.1,
+        },
+        vary = {
+            type = "boolean",
+        }
+    },
+}
+
+
+local _M = {
+    version = 0.1,
+    priority = 996,
+    name = "brotli",
+    schema = schema,
+}
+
+
+function _M.check_schema(conf)
+    return core.schema.check(schema, conf)
+end
+
+
+local function create_brotli_compressor(mode, comp_level, lgwin, lgblock)
+    local options = {
+        mode = mode,
+        quality = comp_level,
+        lgwin = lgwin,
+        lgblock = lgblock,
+    }
+    return brotli.compressor:new(options)
+end
+
+
+local function brotli_compress(conf, ctx, body)
+    local compressor = create_brotli_compressor(conf.mode, conf.comp_level,
+                                                conf.lgwin, conf.lgblock)
+    if not compressor then
+        core.log.error("failed to create brotli compressor")
+        return
+    end
+
+    local chunk = compressor:compress(body)
+    local chunk_fin = compressor:finish()
+    return chunk .. chunk_fin
+end
+
+
+function _M.header_filter(conf, ctx)
+    local types = conf.types
+    local content_type = ngx_header["Content-Type"]
+    if not content_type then
+        -- Like Nginx, don't compress if Content-Type is missing
+        return
+    end
+
+    if type(types) == "table" then
+        local matched = false
+        local from = core.string.find(content_type, ";")
+        if from then
+            content_type = str_sub(content_type, 1, from - 1)
+        end
+
+        for _, ty in ipairs(types) do
+            if content_type == ty then
+                matched = true
+                break
+            end
+        end
+
+        if not matched then
+            return
+        end
+    end
+
+    local content_length = tonumber(ngx_header["Content-Length"])
+    if content_length then
+        local min_length = conf.min_length
+        if content_length < min_length then
+            return
+        end
+        -- Like Nginx, don't check min_length if Content-Length is missing
+    end
+
+    local http_version = req_http_version()
+    if http_version < conf.http_version then
+        return
+    end
+
+    if conf.vary then
+        core.response.add_header("Vary", "Accept-Encoding")
+    end
+
+    ctx.brotli_matched = true
+    core.response.clear_header_as_body_modified()
+    core.response.add_header("Content-Encoding", "br")
+end
+
+
+function _M.body_filter(conf, ctx)
+    if ctx.brotli_matched then
+        local body = core.response.hold_body_chunk(ctx)
+        if ngx.arg[2] == false and not body then
+            return
+        end
+
+        local compressed = brotli_compress(conf, ctx, body)
```

Comment:
What happens if we don't install the `brotli` library?
Is it better to check if the `brotli` is installed before running the plugin?

Reply:
if we don't install the brotli library, when `require("brotli")` will raise `no such file or directory` error,  I think check before running is better, but I'm not sure what should do in this part, any advice or example?

Reply:
what happens if we only install the lua package `brotli-ffi = 0.3-1` and don't install the c library `brotli` ?
It will raise `no such file or directory` too or other error?

Reply:
> what happens if we only install the lua package `brotli-ffi = 0.3-1` and don't install the c library `brotli` ? It will raise `no such file or directory` too or other error?

will raise same error.

Reply:
Maybe we could do this at the top of the file, like this:
```
local brotli = pcall(require, "brotli")
function _M.header_filter(conf, ctx)
  if not brotli then
      core.log.error()
  end
end

function _M.body_filter(conf, ctx)
if not brotli then
    core.log.error()
end
```